adding dynamically_quantize_per_tensor

HDCharles opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #24
* #22
* #21
* #20
* __->__ #18

Summary: largely adapted from vkuzo's pytorch-labs/ao_benchmarks
quant_primitives, this function corrects some small numerical issues
with that function. Specifically you need to explicitly use fp64
intermediates to calculate scale (python float gets autocast to fp32) in
order to match the cuda reference function.

triton graph of dynamically_quantize_per_tensor:
https://gist.github.com/HDCharles/17300b0c0e2cd2e7a3e49d546dc9e19a

Test Plan: python test_quant_primitives.py TestPerTensorQuantization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45757986](https://our.internmc.facebook.com/intern/diff/D45757986)

HDCharles committed:
adding dynamically_quantize_per_tensor

Summary: largely adapted from vkuzo's pytorch-labs/ao_benchmarks
quant_primitives, this function corrects some small numerical issues
with that function.

Test Plan: python test_quant_primitives.py TestPerTensorQuantization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "adding dynamically_quantize_per_tensor"


Summary: largely adapted from vkuzo's pytorch-labs/ao_benchmarks
quant_primitives, this function corrects some small numerical issues
with that function.

Test Plan: python test_quant_primitives.py TestPerTensorQuantization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented:
> largely adapted from vkuzo's pytorch-labs/ao_benchmarks
quant_primitives, this function corrects some small numerical issues
with that function.

can we spell these out?  This is really important to understand and document.

HDCharles committed:
Update on "adding dynamically_quantize_per_tensor"


Summary: largely adapted from vkuzo's pytorch-labs/ao_benchmarks
quant_primitives, this function corrects some small numerical issues
with that function.

Test Plan: python test_quant_primitives.py TestPerTensorQuantization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "adding dynamically_quantize_per_tensor"


Summary: largely adapted from vkuzo's pytorch-labs/ao_benchmarks
quant_primitives, this function corrects some small numerical issues
with that function.

Test Plan: python test_quant_primitives.py TestPerTensorQuantization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "adding dynamically_quantize_per_tensor"


Summary: largely adapted from vkuzo's pytorch-labs/ao_benchmarks
quant_primitives, this function corrects some small numerical issues
with that function. Specifically you need to explicitly use fp64
intermediates to calculate scale (python float gets autocast to fp32) in
order to match the cuda reference function.

triton graph of dynamically_quantize_per_tensor:
https://gist.github.com/HDCharles/17300b0c0e2cd2e7a3e49d546dc9e19a

Test Plan: python test_quant_primitives.py TestPerTensorQuantization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on ao_experimental/quant_primitives.py:
```diff
@@ -2,6 +2,60 @@
+    max_val_pos = torch.max(max_val, torch.zeros_like(max_val))
+
+    # calculate scale and zero point based on min and max
+    # reference: https://github.com/pytorch/pytorch/blob/e779a30d5097714acea011da6a554e43810b5d0e/aten/src/ATen/native/quantized/cpu/QuantUtils.h#L107
```
nit: would be good to add comments about specific precision choices being made here.

HDCharles replied:
done

HDCharles commented:
> 

done

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45757986).

HDCharles committed:
Update on "adding dynamically_quantize_per_tensor"


Summary: largely adapted from vkuzo's pytorch-labs/ao_benchmarks
quant_primitives, this function corrects some small numerical issues
with that function. Specifically you need to explicitly use fp64
intermediates to calculate scale (python float gets autocast to fp32) in
order to match the cuda reference function.

triton graph of dynamically_quantize_per_tensor:
https://gist.github.com/HDCharles/17300b0c0e2cd2e7a3e49d546dc9e19a

Test Plan: python test_quant_primitives.py TestPerTensorQuantization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45757986](https://our.internmc.facebook.com/intern/diff/D45757986)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45757986).

HDCharles committed:
Update on "adding dynamically_quantize_per_tensor"


Summary: largely adapted from vkuzo's pytorch-labs/ao_benchmarks
quant_primitives, this function corrects some small numerical issues
with that function. Specifically you need to explicitly use fp64
intermediates to calculate scale (python float gets autocast to fp32) in
order to match the cuda reference function.

triton graph of dynamically_quantize_per_tensor:
https://gist.github.com/HDCharles/17300b0c0e2cd2e7a3e49d546dc9e19a

Test Plan: python test_quant_primitives.py TestPerTensorQuantization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45757986](https://our.internmc.facebook.com/intern/diff/D45757986)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45757986).

HDCharles committed:
Update on "adding dynamically_quantize_per_tensor"


Summary: largely adapted from vkuzo's pytorch-labs/ao_benchmarks
quant_primitives, this function corrects some small numerical issues
with that function. Specifically you need to explicitly use fp64
intermediates to calculate scale (python float gets autocast to fp32) in
order to match the cuda reference function.

triton graph of dynamically_quantize_per_tensor:
https://gist.github.com/HDCharles/17300b0c0e2cd2e7a3e49d546dc9e19a

Test Plan: python test_quant_primitives.py TestPerTensorQuantization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45757986](https://our.internmc.facebook.com/intern/diff/D45757986)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45757986).

HDCharles committed:
Update on "adding dynamically_quantize_per_tensor"


Summary: largely adapted from vkuzo's pytorch-labs/ao_benchmarks
quant_primitives, this function corrects some small numerical issues
with that function. Specifically you need to explicitly use fp64
intermediates to calculate scale (python float gets autocast to fp32) in
order to match the cuda reference function.

triton graph of dynamically_quantize_per_tensor:
https://gist.github.com/HDCharles/17300b0c0e2cd2e7a3e49d546dc9e19a

Test Plan: python test_quant_primitives.py TestPerTensorQuantization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45757986](https://our.internmc.facebook.com/intern/diff/D45757986)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45757986).

HDCharles committed:
Update on "adding dynamically_quantize_per_tensor"


Summary: largely adapted from vkuzo's pytorch-labs/ao_benchmarks
quant_primitives, this function corrects some small numerical issues
with that function. Specifically you need to explicitly use fp64
intermediates to calculate scale (python float gets autocast to fp32) in
order to match the cuda reference function.

triton graph of dynamically_quantize_per_tensor:
https://gist.github.com/HDCharles/17300b0c0e2cd2e7a3e49d546dc9e19a

Test Plan: python test_quant_primitives.py TestPerTensorQuantization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45757986](https://our.internmc.facebook.com/intern/diff/D45757986)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45757986).
